File: Image2LinearGradientCSS.js
/**
 * Image2LinearGradientCSS.js
 *
 * A proof-of-concept to convert an image to linear-gradient CSS
 * usage: node Image2LinearGradientCSS.js "input url/filepath" "output txt file"
 * example: node Image2LinearGradientCSS.js "https://raw.githubusercontent.com/Maxoplata/Image2LinearGradientCSS/main/example/example.jpg" "./example.txt"
 * example: node Image2LinearGradientCSS.js "./example.jpg" "./example.txt"
 *
 * @author Maxamilian Demian
 * @link https://www.maxodev.org
 * @link https://github.com/Maxoplata/Image2LinearGradientCSS
 */

const fs = require('node:fs/promises');
const Jimp = require('jimp');

(async () => {
	if (process.argv.length != 4) {
		throw new Error('Invalid argument count');
	}

    // load our command line params
	const inputFile = process.argv[2];
	const outputFile = process.argv[3];

	try {
		// load input file
		const image = await Jimp.read(inputFile);

        // our output css
        let output = '';

        // keep track of the pixels we have already accounted for
        let visited = Array.from({ length: image.bitmap.height }, () => Array(image.bitmap.width).fill(false));

        for (let y = 0; y < image.bitmap.height; y++) {
            for (let x = 0; x < image.bitmap.width; x++) {
                // skip already processed pixels
                if (visited[y][x]) continue;

                // get pixel color
                let color = image.getPixelColor(x, y);

                // determine the width and height of the color block
                let blockWidth = 1;
                let blockHeight = 1;

                // calculate block width
                while (x + blockWidth < image.bitmap.width && image.getPixelColor(x + blockWidth, y) === color) {
                    blockWidth++;
                }

                // calculate block height
                while (y + blockHeight < image.bitmap.height && image.getPixelColor(x, y + blockHeight) === color) {
                    let isSameColor = true;

                    for (let i = 0; i < blockWidth; i++) {
                        if (image.getPixelColor(x + i, y + blockHeight) !== color) {
                            isSameColor = false;

                            break;
                        }
                    }

                    if (!isSameColor) break;

                    blockHeight++;
                }

                // mark the pixels in this block as visited
                for (let by = 0; by < blockHeight; by++) {
                    for (let bx = 0; bx < blockWidth; bx++) {
                        visited[y + by][x + bx] = true;
                    }
                }

                // add the block to the output
                output += (output !== '' ? ', ' : '') + `linear-gradient(#${color.toString(16).substring(0, 6).toUpperCase()} 100%, white 100%) no-repeat scroll ${x}px ${y}px / ${blockWidth}px ${blockHeight}px`;
			}
		}

		// write output to file
		await fs.writeFile(outputFile, `background: ${output};`);

		console.log(`background CSS for '${inputFile}'' has been written to '${outputFile}'`);
	} catch (error) {
		throw new Error(`Failed to create background CSS output: ${error}`);
	}
})();

File: Image2LinearGradientCSS.rb
#!/usr/bin/ruby
# Image2LinearGradientCSS.rb
#
# A proof-of-concept to convert an image to linear-gradient CSS
# usage: ruby Image2LinearGradientCSS.rb "input url/filepath" "output txt file"
# example: ruby Image2LinearGradientCSS.rb "https://raw.githubusercontent.com/Maxoplata/Image2LinearGradientCSS/main/example/example.jpg" "./example.txt"
# example: ruby Image2LinearGradientCSS.rb "./example.jpg" "./example.txt"
#
# @author Maxamilian Demian
# @link https://www.maxodev.org
# @link https://github.com/Maxoplata/Image2LinearGradientCSS
require 'open-uri'
require 'rmagick'

def rgbToHex(rgb)
    r = (rgb.red / 257).to_i
    g = (rgb.green / 257).to_i
    b = (rgb.blue / 257).to_i

    sprintf("%02X%02X%02X", r, g, b)
end

if ARGV.count != 2
	raise 'Invalid argument count'
end

# load our command line params
inputFile = ARGV[0]
outputFile = ARGV[1]

begin
    image = nil

    # load input file
    if File.file?(inputFile)
        image = Magick::Image.read(inputFile).first
    else
        # Magick::Image.read can handle URLs, but fails on HTTPS
        # so we manually handle URLs
        urlData = URI.open(inputFile).read

        image = Magick::ImageList.new

        image.from_blob(urlData)
    end

	# get width/height of image
	imageWidth = image.columns
	imageHeight = image.rows

    # our output css
    output = ''

    # keep track of the pixels we have already accounted for
    visited = Array.new(imageHeight) { Array.new(imageWidth, false) }

	for y in 0..(imageHeight - 1) do
        for x in 0..(imageWidth - 1) do
            # skip already processed pixels
            next if visited[y][x]

            # get pixel color
			color = rgbToHex(image.pixel_color(x, y))

            # determine the width and height of the color block
            blockWidth = 1
            blockHeight = 1

            # calculate block width
            while x + blockWidth < imageWidth && rgbToHex(image.pixel_color(x + blockWidth, y)) == color do
                blockWidth += 1
            end

            # calculate block height
            while y + blockHeight < imageHeight && rgbToHex(image.pixel_color(x, y + blockHeight)) == color do
                isSameColor = true

                for i in 0..(blockWidth - 1) do
                    if rgbToHex(image.pixel_color(x + i, y + blockHeight)) != color
                        isSameColor = false

                        break
                    end
                end

                break if !isSameColor

                blockHeight += 1
            end

            # mark the pixels in this block as visited
            for by in 0..(blockHeight - 1) do
                for bx in 0..(blockWidth - 1) do
                    visited[y + by][x + bx] = true
                end
            end

            # add the block to the output
            output += (output != '' ? ', ' : '') + "linear-gradient(##{color} 100%, white 100%) no-repeat scroll #{x}px #{y}px / #{blockWidth}px #{blockHeight}px"
        end
    end

    # write output to file
    File.write(outputFile, "background: #{output};")

    puts "background CSS for '#{inputFile}' has been written to '#{outputFile}'"
rescue => error
    puts "Failed to create background CSS output: #{error.message}"
end
